tsl/src/fdw/deparse.c

```diff
@@ -834,6 +834,169 @@ deparseSelectStmtForRel(StringInfo buf, PlannerInfo *root, RelOptInfo *rel, List
 	deparseLockingClause(&context);
 }
 
+/*
+ * Construct "SELECT DISTINCT target_list" or "SELECT DISTINCT ON (col1, col..)
+ * target_list" statement to push down the DISTINCT clause to the remote side.
+ *
+ * We only allow references to basic "Vars" or constants in the DISTINCT exprs
+ *
+ * So, "SELECT DISTINCT col1" is fine but "SELECT DISTINCT 2*col1" is not.
+ *
+ * "SELECT DISTINCT col1, 'const1', NULL, col2" which is a mix of column
+ * references and constants is also supported. Everything else is not supported.
+ *
+ * Note that "SELECT DISTINCT col1, col2" is different from "SELECT DISTINCT
+ * col2, col1". So, the attributes have to be fetched in the same order as
+ * well from the remote side. This means that "retrieved_attrs" list also
+ * needs to be constructed in the same order. This function returns a
+ * "dc_varattnos" list which points to the attributes as per the DISTINCT
+ * specification.
+ */
+static void
+deparseDistinctClause(StringInfo buf, List **dc_varattnos, deparse_expr_cxt *context)
+{
+	PlannerInfo *root = context->root;
+	Query *query = root->parse;
+	ListCell *l, *dc_l;
+	bool first = true;
+	bool dc_pushdown_safe = true;
+	bool have_wholerow = false;
+	Index varno;
+	/*
+	 * Sortgroups to deparse for SELECT DISTINCT. Only "Vars" will go in
+	 * this list. Constants will not go
+	 */
+	List *sgc_deparse_list = NIL;
+
+	*dc_varattnos = NIL;
+
+	if (query->distinctClause == NIL)
+		return;
+
+	foreach (l, query->distinctClause)
+	{
+		SortGroupClause *sgc = (SortGroupClause *) lfirst(l);
+		TargetEntry *tle = get_sortgroupclause_tle(sgc, query->targetList);
+
+		/*
+		 * We only send basic attributes to the remote side. So we can
+		 * pushdown DISTINCT only if the tle is a simple one
+		 * referring to the "Var" directly. Also all varno entries
+		 * need to point to the same relid.
+		 *
+		 * Collect the varattnos in the below loop since it's important
+		 * to print the targetlist in the DISTINCT order. This is
+		 * necessary because "DISTINCT col2, col1" is very different
+		 * from "DISTINCT col1, col2"
+		 *
+		 * Also handle "DISTINCT col1, CONST1, NULL" types cases
+		 */
+		if (IsA(tle->expr, Var))
+		{
+			Var *var = (Var *) tle->expr;
+
+			if (first)
+			{
+				varno = var->varno;
+				first = false;
+			}
+
+			if (varno != var->varno)
+				dc_pushdown_safe = false;
```

Comment:
It seems the pushdown_fase=false case is never really tested. Is it possible to add a test for this?

Reply:
Ok, added a test case containing an expression on columns. Expressions don't get pushed down. Only plain Vars do.